packages/web/app/src/lib/fetch-json.ts

```diff
@@ -0,0 +1,134 @@
+> {
+  const response = await fetch(url, requestInit)
+    // @see https://developer.mozilla.org/en-US/docs/Web/API/Window/fetch#exceptions
+    .catch(Kit.oneOf(error => error instanceof TypeError || error instanceof DOMException));
+
+  if (response instanceof TypeError) {
+    return new FetchJsonErrors.FetchJsonRequestTypeError({ requestInit }, response);
+  }
+
+  if (response instanceof DOMException) {
+    return new FetchJsonErrors.FetchJsonRequestNetworkError({}, response);
+  }
+
+  const json = await response
+    .json()
+    .then(value => value as Kit.Json.Value)
+    // @see https://developer.mozilla.org/en-US/docs/Web/API/Response/json#exceptions
+    .catch(
+      Kit.oneOf(
+        error =>
+          error instanceof SyntaxError ||
+          error instanceof TypeError ||
+          error instanceof DOMException,
+      ),
+    );
+
+  if (json instanceof DOMException) {
+    return new FetchJsonErrors.FetchJsonRequestNetworkError({}, json);
+  }
+
+  if (json instanceof TypeError) {
+    return new FetchJsonErrors.FetchJsonResponseTypeError({ response }, json);
+  }
+
+  if (json instanceof SyntaxError) {
+    return new FetchJsonErrors.FetchJsonResponseSyntaxError({ response }, json);
+  }
+
+  if (schema) {
+    const result = schema.safeParse(json);
+    if (!result.success) {
+      return new FetchJsonErrors.FetchJsonResponseSchemaError(
+        { response, json, schema },
+        result.error,
+      );
+    }
+    return result.data as any; // z.infer<Exclude<schema, undefined>>;
+  }
+
+  return json as any; // Kit.Json.Value;
+}
+
+// =================================
+// Error Classes
+// =================================
+
+// eslint-disable-next-line @typescript-eslint/no-namespace
+export namespace FetchJsonErrors {
+  export type FetchJsonErrors = FetchJsonResponseErrors | FetchJsonRequestErrors;
+
+  // --------------------------------
+  // Response Error Classes
+  // --------------------------------
+
+  export type FetchJsonRequestErrors = FetchJsonRequestTypeError | FetchJsonRequestNetworkError;
+
+  export class FetchJsonRequestNetworkError extends Kit.Errors.ContextualError<
+    'FetchJsonRequestNetworkError',
+    {},
+    DOMException
+  > {
+    message = 'Network failure.';
+  }
+
+  export class FetchJsonRequestTypeError extends Kit.Errors.ContextualError<
+    'FetchJsonRequestTypeError',
+    { requestInit?: RequestInit },
+    TypeError
+  > {
+    message = 'Invalid request.';
+  }
+
+  // --------------------------------
+  // Response Error Classes
+  // --------------------------------
+
+  export abstract class FetchJsonResponseError<
+    $Name extends string,
+    $Context extends {
+      response: Response;
+    },
+    $Cause extends z.ZodError | SyntaxError | TypeError | DOMException,
+  > extends Kit.Errors.ContextualError<$Name, $Context, $Cause> {
+    message = 'Invalid response.';
```

Comment:
Prefer if we can provide more insight than this. E.g. should we include the cause's error message?

Reply:
It's all available through the chain data in a structured way. The missing part IMO is error renderers that are up to date with modern error types. Cause is part of the ECMA  standard so it's confusing to me that default renderings are still AFAIK failing to raise good context about them. We would just have to add our own where it is needed such as top level catch all or logging middleware, etc.